tools/wake-format/formatter.h

```diff
@@ -238,6 +303,12 @@ struct Formatter {
     assert(node.empty());
     return std::move(builder).build();
   }
+
+  wcl::doc compose(ctx_t ctx, CSTElement& node) {
```

Comment:
Is this just to avoid a friend function?

Reply:
This shares the node reference with the composed formatter. I don't think its sufficient to make a friend function because you have to advance the node reference of the `parent` class and you can't just do simple assignment since the formatter class doesn't actually store its node, instead it only manipulates the reference created from the top level format call.

(that said, I do think minimizing friend classes is a preferable thing to do)